### PR TITLE
[DEV-140033] retreive secrets based on dotenv files with sm:// paths

### PR DESCRIPTION
**Pull request by tonitienda-rovtech**
## Summary
Adds support for `.env` file sources with automatic AWS Secrets Manager resolution. Users can now reference secrets directly in env files using `sm://` paths, and snagsby will automatically fetch and inject the secret values.

### New Feature: EnvFileResolver
- Added `EnvFileResolver` to support `file://` scheme sources
- Parses standard `.env` file format (KEY=value pairs)
- Automatically resolves values prefixed with `sm://` from AWS Secrets Manager
- Supports comments (`#`) and inline comments
- Preserves original file order in output

### Usage Example
```bash
# Create an env file with secret references
cat > app.env << 'HEREDOC'
DATABASE_URL=sm://prod/db/connection-string
API_KEY=sm://prod/api/key
LOG_LEVEL=info
HEREDOC

# Snagsby will fetch secrets and export all values
./snagsby "file://./app.env?region=us-west-2"
```

## Code Quality Improvements (from PR review)

Following TDD methodology, added tests first to demonstrate issues, then fixed the code:

### 1. **Empty Key Validation**
- **Issue**: Lines like `=value` were silently skipped without error
- **Fix**: Now returns error for malformed lines with empty keys
- **Location**: `pkg/resolvers/envfile.go:47-49`

### 2. **Duplicate Key Detection with Normalization**
- **Issue**: Keys like `foo-bar` and `FOO_BAR` weren't caught as duplicates, causing silent overwrites
- **Fix**: Normalize keys (uppercase + replace non-word chars) before duplicate checking to match final output
- **Location**: `pkg/resolvers/envfile.go:54-83`

### 3. **Secret Key Deduplication**
- **Issue**: Multiple env vars pointing to same `sm://` path caused redundant API calls
- **Fix**: Deduplicate secret paths before calling GetSecrets to reduce AWS API load
- **Location**: `pkg/resolvers/envfile.go:117-120`

### 4. **Relative File Path Support**
- **Issue**: Relative paths like `file://./pre-cache.env` were incorrectly parsed
- **Fix**: Properly handle URL parsing for relative paths (`.`, `..`, simple filenames)
- **Location**: `pkg/resolvers/envfile.go:27-38`

### 5. **Test Coverage**
- Added assertions for `expectedSecretsRequested` to verify secret resolution behavior
- Added `file://` scheme test to `TestResolveSource`
- Added comprehensive tests for edge cases (empty keys, normalization, deduplication)

## Testing
All changes follow TDD - tests written first to fail, then code fixed:
```bash
go test ./pkg/resolvers/...
# All 40+ tests pass
```

## Files Changed
- `pkg/resolvers/envfile.go` - New resolver implementation
- `pkg/resolvers/envfile_test.go` - Comprehensive test coverage
- `pkg/resolvers/resolvers.go` - Added file scheme routing
- `pkg/resolvers/resolvers_test.go` - Added file scheme test
- `pkg/connectors/secretsmanager.go` - Exposed GetSecrets method
- `pkg/connectors/testing/mocks.go` - Added mock connector for testing

## Related
Closes DEV-140033

### CODE REVIEW

File: pkg/resolvers/resolvers.go [hunk 1]
```diff
@@ -88,6 +88,12 @@ func ResolveSource(source *config.Source) *Result {
 			return &Result{Source: source, Errors: []error{err}}
 		}
 		s = NewManifestResolver(connector)
+	case "file":
+		connector, err := connectors.NewSecretsManagerConnector(source)
+		if err != nil {
+			return &Result{Source: source, Errors: []error{err}}
+		}
+		s = NewEnvFileResolver(connector)
```
[the remark this answers is not in the document]
Comment (tonitienda-rovtech) [reply pkg/resolvers/resolvers.go:119]: I will not implement this change.
I would argue that if the env file has no references to sm:// you would not need to use snagsby at all.

File: pkg/resolvers/envfile.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+			return source.URL.Host
+		}
+		// Remove leading slash from path to avoid double slashes
+		path := strings.TrimPrefix(source.URL.Path, "/")
+		return fmt.Sprintf("%s/%s", source.URL.Host, path)
+	}
+	return source.URL.Path
+}
+
+func processLine(line string) (string, string, error) {
```
Comment (eimhin-rover) [inline pkg/resolvers/envfile.go]: Can you give a bit more context on what the purpose of this function is?

It also seems like it is explicitly used by the EnvFileReoslver so maybe it belongs as a function declared on the struct?

File: pkg/resolvers/envfile.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		secretKeysMap[secretPath] = true
+	}
+	secretKeys := slices.Collect(maps.Keys(secretKeysMap))
+	secrets, errors := e.connector.GetSecrets(secretKeys)
+
+	for _, err := range errors {
+		result.AppendError(err)
+	}
+
+	for _, key := range envVarsOrder {
```
Comment (eimhin-rover) [inline pkg/resolvers/envfile.go]: This is fairly deep nesting which is a little difficult to read.

File: pkg/resolvers/envfile.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+}
+
+func isValidEnvVarName(key string) bool {
+	return envVarNameRegexp.MatchString(key)
+}
+
+func getFilePath(source *config.Source) string {
+	// For file:// URLs, if there's a host component (e.g., file://./path or file://filename),
+	// we need to concatenate host and path to preserve relative paths
+	if source.URL.Host != "" {
```
Comment (eimhin-rover) [inline pkg/resolvers/envfile.go]: There's a risk of a nil pointer exception here since you aren't checking the value of `source`

**Review by eimhin-rover — CHANGES_REQUESTED** [review]

File: pkg/resolvers/envfile.go (hunk 4 above)
Comment (tonitienda-rovtech) [reply pkg/resolvers/envfile.go]: I am adding an assertion at the entry point (resolvers.go).

File: pkg/resolvers/envfile.go (hunk 2 above)
Comment (tonitienda-rovtech) [reply pkg/resolvers/envfile.go]: I am adding a bit of doc and improving the function name.

I prefer to keep is as a pure/standalone function because it does not need any of the struct dependencies or read/change its state.

File: pkg/resolvers/envfile.go (hunk 3 above)
Comment (tonitienda-rovtech) [reply pkg/resolvers/envfile.go]: Refactoring the function in clear sections/subfunctions.

File: pkg/resolvers/envfile.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	if key == "" {
+		return "", "", fmt.Errorf("invalid line: %s (empty key)", line)
+	}
+
+	if !isValidEnvVarName(key) {
+		return "", "", fmt.Errorf("invalid key '%s': environment variable names must contain only letters, digits, and underscores, and must start with a letter or underscore", key)
+	}
+
+	// Remove inline comments
+	if idx := strings.Index(value, " #"); idx != -1 {
```
Comment (JVenberg) [inline pkg/resolvers/envfile.go:100]: Are you sure this doesn't remove # values from the secrets themselves? Do you have any test cases that would validate that behavior?

Comment (tonitienda-rovtech) [reply pkg/resolvers/envfile.go:100]: I added more tests and the code was fixed

File: pkg/resolvers/envfile.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+	// Handle quoted values (both single and double quotes)
+	if len(value) >= 1 {
+		firstChar := value[0]
+		if firstChar == '"' || firstChar == '\'' {
+			// Find the closing quote
+			closingQuoteIdx := strings.IndexByte(value[1:], firstChar)
+			if closingQuoteIdx == -1 {
+				// No closing quote found
+				return key, "", fmt.Errorf("invalid line: %s (uneven quotes)", line)
+			}
```
[the remark this answers is not in the document]
Comment (tonitienda-rovtech) [reply pkg/resolvers/envfile.go:89]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/roverdotcom/snagsby/pull/53#discussion_r2883213706)

**Review by eimhin-rover — APPROVED** [review]

File: README.md [hunk 7]
```diff
@@ -1,10 +1,15 @@
...
-in a format that can be evaluated by a shell to set environment variables.
+Snagsby reads configuration and secrets from multiple sources and outputs them
+as environment variables in a format that can be evaluated by a shell.
 
-It's useful for reading configuration into environment variables from S3
-objects in Docker containers.
+**Supported sources:**
+- Local env files (`file://`) with dotenv format
+- AWS S3 JSON objects (`s3://`)
+- AWS Secrets Manager (`sm://`)
```
[the remark this answers is not in the document]
Comment (tonitienda-rovtech) [reply README.md:9]: It was not adverstised in the past and we think it is not being used. We plan to deprecate it soon so we prefer not to advertise it now,
